<a href="https://colab.research.google.com/github/aruaru0/CommonLit-Evaluate-Student-Summaries/blob/main/tuned_debertav3L_lgbm_autocorrect.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import sys
is_colab = False
if "google.colab" in sys.modules:
  is_colab = True

In [2]:
if is_colab:
  from google.colab import drive
  drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
if is_colab:
  !mkdir -p /root/.kaggle
  !cp /content/drive/MyDrive/Kaggle/kaggle.json  /root/.kaggle/
  !chmod 600 /root/.kaggle/kaggle.json

In [4]:
if is_colab:
  !pip install kaggle
  !apt install unzip
  !mkdir input output

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
unzip is already the newest version (6.0-26ubuntu3.1).
0 upgraded, 0 newly installed, 0 to remove and 18 not upgraded.


In [5]:
if is_colab:
  !kaggle competitions download -c commonlit-evaluate-student-summaries
  !unzip -o commonlit-evaluate-student-summaries.zip -d input/commonlit-evaluate-student-summaries

 95% 1.00M/1.05M [00:00<00:00, 1.99MB/s]
100% 1.05M/1.05M [00:00<00:00, 2.08MB/s]
Archive:  commonlit-evaluate-student-summaries.zip
  inflating: input/commonlit-evaluate-student-summaries/prompts_test.csv  
  inflating: input/commonlit-evaluate-student-summaries/prompts_train.csv  
  inflating: input/commonlit-evaluate-student-summaries/sample_submission.csv  
  inflating: input/commonlit-evaluate-student-summaries/summaries_test.csv  
  inflating: input/commonlit-evaluate-student-summaries/summaries_train.csv  


In [ ]:
if is_colab:
  !pip install transformers
  !pip install datasets
  !pip install sentencepiece
  !pip install autocorrect==2.6
  !pip install pyspellchecker==0.7.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 24.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.9/294.9 kB 26.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 89.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 79.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.6/519.6 kB 5.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 10.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 11.0 MB/s eta 0:00:00


In [ ]:
if is_colab:
  !pip install git+https://github.com/huggingface/accelerate

# ここでランタイムの再起動が必要！！

In [ ]:
import sys
is_colab = False
if "google.colab" in sys.modules:
  is_colab = True

if is_colab:
  %cd output

/content/output


In [ ]:
if is_colab:
  from transformers import AutoModel, AutoTokenizer, AutoConfig, AutoModelForSequenceClassification
  model_name = "microsoft/deberta-v3-large"
  tokenizer = AutoTokenizer.from_pretrained(model_name)
  hidden_dropout_prob=0.007
  attention_probs_dropout_prob=0.007
  model = AutoModelForSequenceClassification.from_pretrained(model_name)
  model_dir = "../input/another-bert"
  model.save_pretrained(model_dir)
  tokenizer.save_pretrained(model_dir)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
/usr/local/lib/python3.10/dist-packages/transformers/convert_slow_tokenizer.py:473: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  warnings.warn(
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at microsoft/deberta-v3-large and are newly initialized: ['classifier.bias', 'pooler.dense.weight', 'classifier.weight', 'pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
if is_colab:
  import nltk
  nltk.download('stopwords')
  nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


# Notes
This is a slightly tuned version of @nogawanogawa 's work and I have also converted his messages to english here you can find his notebook here https://www.kaggle.com/code/tsunotsuno/updated-debertav3-lgbm-with-spell-autocorrect please give him kudos for sharing his efforts

### Things I would expect there to be a number of things that will allow this model to preform better outside of just strategy and more data. I would imagine there are a few more tuning parameters that could help this model go a long way.





In this notebook a combonation of Deberta and LGBM is used, pyspellchecker is also used in order to correct some of the spelling mistakes that are discussed in the discussions tab
[Discussion Link](https://www.kaggle.com/competitions/commonlit-evaluate-student-summaries/discussion/428941).
[my previous notebook](https://www.kaggle.com/code/tsunotsuno/debertav3-lgbm-with-feature-engineering)
[Discussion Link](https://www.kaggle.com/competitions/commonlit-evaluate-student-summaries/discussion/428941).

The primary goal of this notebook is to enhance the overall score by honing in on the issue of "misspellings."

## Main Concept

The Transformers model I'm currently utilizing, Deberta, is pretrained on "correct sentences." However, if I were to train and input it with sentences containing misspellings, Deberta's ability to understand meaning might be compromised.

From a human evaluator's perspective, detecting misspellings would prompt deductions in scores. After discreetly rectifying the misspelled words, I'd proceed to evaluate other textual facets. If we assume the scoring process aligns with this approach, it's conceivable that tallying and **correcting** misspellings before feeding text into Deberta could enable the model to aptly capture features beyond just misspellings.

In this notebook, I will embark on the journey of auto-correcting misspelled words before inputting them into Deberta. The aim is to evaluate the model's performance by distinctly isolating misspellings from other aspects.

### Feature Engineering

I intend to largely retain the same features as before:

- Text Length
- Length Ratio
- Word Overlap
- N-grams Co-occurrence
  - Count
  - Ratio
- Quotes Overlap
- Grammar Check
  - Spelling: pyspellchecker

### Model Architecture

I plan to construct a model with the architecture depicted in the following diagram. For the input to Deberta (`text`), I will pre-process by correcting any misspellings. In other aspects of feature engineering, I will utilize the `text` as is.

### References

- https://www.kaggle.com/competitions/commonlit-evaluate-student-summaries/discussion/428941

### My previous notebooks

- https://www.kaggle.com/code/tsunotsuno/debertav3-baseline-content-and-wording-models
- https://www.kaggle.com/code/tsunotsuno/debertav3-w-prompt-title-question-fields
- https://www.kaggle.com/code/tsunotsuno/debertav3-with-llama2-example
- https://www.kaggle.com/code/tsunotsuno/debertav3-lgbm-with-feature-engineering


In [ ]:
if not is_colab :
  !pip install "/kaggle/input/autocorrect/autocorrect-2.6.1.tar"
  !pip install "/kaggle/input/pyspellchecker/pyspellchecker-0.7.2-py3-none-any.whl"

In [ ]:
from typing import List
import numpy as np
import pandas as pd
import warnings
import logging
import os
import shutil
import json
import transformers
from transformers import AutoModel, AutoTokenizer, AutoConfig, AutoModelForSequenceClassification
from transformers import DataCollatorWithPadding
from datasets import Dataset,load_dataset, load_from_disk
from transformers import TrainingArguments, Trainer
from datasets import load_metric, disable_progress_bar
from sklearn.metrics import mean_squared_error
import torch
from sklearn.model_selection import KFold, GroupKFold
from tqdm import tqdm

import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.tokenize.treebank import TreebankWordDetokenizer
from collections import Counter
import spacy
import re
from autocorrect import Speller
from spellchecker import SpellChecker
import lightgbm as lgb

warnings.simplefilter("ignore")
logging.disable(logging.ERROR)
os.environ["TOKENIZERS_PARALLELISM"] = "false"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
disable_progress_bar()
tqdm.pandas()

In [ ]:
def seed_everything(seed: int):
    import random, os
    import numpy as np
    import torch

    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True

seed_everything(seed=42)

In [ ]:
class CFG:
    model_name="another-bert"
    learning_rate=1.5e-5
    weight_decay=0.02
    hidden_dropout_prob=0.007
    attention_probs_dropout_prob=0.007
    num_train_epochs=5
    n_splits=4
    batch_size=3

    random_seed=42
    save_steps=100
    max_length=512

## Dataload

In [ ]:
DATA_DIR = "../input/commonlit-evaluate-student-summaries/"

prompts_train = pd.read_csv(DATA_DIR + "prompts_train.csv")
prompts_test = pd.read_csv(DATA_DIR + "prompts_test.csv")
summaries_train = pd.read_csv(DATA_DIR + "summaries_train.csv")
summaries_test = pd.read_csv(DATA_DIR + "summaries_test.csv")
sample_submission = pd.read_csv(DATA_DIR + "sample_submission.csv")

## Preprocess

[Using features]

- Text Length
- Length Ratio
- Word Overlap
- N-grams Co-occurrence
  - count
  - ratio
- Quotes Overlap
- Grammar Check
  - spelling: pyspellchecker


In [ ]:
class Preprocessor:
    def __init__(self,
                model_name: str,
                ) -> None:
        self.tokenizer = AutoTokenizer.from_pretrained(f"../input/{model_name}")
        self.twd = TreebankWordDetokenizer()
        self.STOP_WORDS = set(stopwords.words('english'))

        self.spacy_ner_model = spacy.load('en_core_web_sm',)
        self.speller = Speller(lang='en')
        self.spellchecker = SpellChecker()

    def word_overlap_count(self, row):
        """ intersection(prompt_text, text) """
        def check_is_stop_word(word):
            return word in self.STOP_WORDS

        prompt_words = row['prompt_tokens']
        summary_words = row['summary_tokens']
        if self.STOP_WORDS:
            prompt_words = list(filter(check_is_stop_word, prompt_words))
            summary_words = list(filter(check_is_stop_word, summary_words))
        return len(set(prompt_words).intersection(set(summary_words)))

    def ngrams(self, token, n):
        # Use the zip function to help us generate n-grams
        # Concatentate the tokens into ngrams and return
        ngrams = zip(*[token[i:] for i in range(n)])
        return [" ".join(ngram) for ngram in ngrams]

    def ngram_co_occurrence(self, row, n: int) -> int:
        # Tokenize the original text and summary into words
        original_tokens = row['prompt_tokens']
        summary_tokens = row['summary_tokens']

        # Generate n-grams for the original text and summary
        original_ngrams = set(self.ngrams(original_tokens, n))
        summary_ngrams = set(self.ngrams(summary_tokens, n))

        # Calculate the number of common n-grams
        common_ngrams = original_ngrams.intersection(summary_ngrams)
        return len(common_ngrams)

    def ner_overlap_count(self, row, mode:str):
        model = self.spacy_ner_model
        def clean_ners(ner_list):
            return set([(ner[0].lower(), ner[1]) for ner in ner_list])
        prompt = model(row['prompt_text'])
        summary = model(row['text'])

        if "spacy" in str(model):
            prompt_ner = set([(token.text, token.label_) for token in prompt.ents])
            summary_ner = set([(token.text, token.label_) for token in summary.ents])
        elif "stanza" in str(model):
            prompt_ner = set([(token.text, token.type) for token in prompt.ents])
            summary_ner = set([(token.text, token.type) for token in summary.ents])
        else:
            raise Exception("Model not supported")

        prompt_ner = clean_ners(prompt_ner)
        summary_ner = clean_ners(summary_ner)

        intersecting_ners = prompt_ner.intersection(summary_ner)

        ner_dict = dict(Counter([ner[1] for ner in intersecting_ners]))

        if mode == "train":
            return ner_dict
        elif mode == "test":
            return {key: ner_dict.get(key) for key in self.ner_keys}


    def quotes_count(self, row):
        summary = row['text']
        text = row['prompt_text']
        quotes_from_summary = re.findall(r'"([^"]*)"', summary)
        if len(quotes_from_summary)>0:
            return [quote in text for quote in quotes_from_summary].count(True)
        else:
            return 0

    def spelling(self, text):

        wordlist=text.split()
        amount_miss = len(list(self.spellchecker.unknown(wordlist)))

        return amount_miss

    def add_spelling_dictionary(self, tokens: List[str]) -> List[str]:
        """dictionary update for pyspell checker and autocorrect"""
        self.spellchecker.word_frequency.load_words(tokens)
        self.speller.nlp_data.update({token:1000 for token in tokens})

    def run(self,
            prompts: pd.DataFrame,
            summaries:pd.DataFrame,
            mode:str
        ) -> pd.DataFrame:

        # before merge preprocess
        prompts["prompt_length"] = prompts["prompt_text"].apply(
            lambda x: len(word_tokenize(x))
        )
        prompts["prompt_tokens"] = prompts["prompt_text"].apply(
            lambda x: word_tokenize(x)
        )

        summaries["summary_length"] = summaries["text"].apply(
            lambda x: len(word_tokenize(x))
        )
        summaries["summary_tokens"] = summaries["text"].apply(
            lambda x: word_tokenize(x)
        )

        # Add prompt tokens into spelling checker dictionary
        prompts["prompt_tokens"].apply(
            lambda x: self.add_spelling_dictionary(x)
        )

#         from IPython.core.debugger import Pdb; Pdb().set_trace()
        # fix misspelling
        summaries["fixed_summary_text"] = summaries["text"].progress_apply(
            lambda x: self.speller(x)
        )

        # count misspelling
        summaries["splling_err_num"] = summaries["text"].progress_apply(self.spelling)

        # merge prompts and summaries
        input_df = summaries.merge(prompts, how="left", on="prompt_id")

        # after merge preprocess
        input_df['length_ratio'] = input_df['summary_length'] / input_df['prompt_length']

        input_df['word_overlap_count'] = input_df.progress_apply(self.word_overlap_count, axis=1)
        input_df['bigram_overlap_count'] = input_df.progress_apply(
            self.ngram_co_occurrence,args=(2,), axis=1
        )
        input_df['bigram_overlap_ratio'] = input_df['bigram_overlap_count'] / (input_df['summary_length'] - 1)

        input_df['trigram_overlap_count'] = input_df.progress_apply(
            self.ngram_co_occurrence, args=(3,), axis=1
        )
        input_df['trigram_overlap_ratio'] = input_df['trigram_overlap_count'] / (input_df['summary_length'] - 2)

        input_df['quotes_count'] = input_df.progress_apply(self.quotes_count, axis=1)

        return input_df.drop(columns=["summary_tokens", "prompt_tokens"])

preprocessor = Preprocessor(model_name=CFG.model_name)

In [ ]:
train = preprocessor.run(prompts_train, summaries_train, mode="train")
test = preprocessor.run(prompts_test, summaries_test, mode="test")

train.head()

100%|██████████| 4/4 [00:00<00:00, 1395.43it/s]


student_id prompt_id                                               text  \
0  000e8c3c7ddb    814d6b  The third wave was an experimentto see how peo...   
1  0020ae56ffbf    ebad26  They would rub it up with soda to make the sme...   
2  004e978e639e    3b9047  In Egypt, there were many occupations and soci...   
3  005ab0199905    3b9047  The highest class was Pharaohs these people we...   
4  0070c9e7af47    814d6b  The Third Wave developed  rapidly because the ...   

    content   wording  summary_length  \
0  0.205683  0.380538              64   
1 -0.548304  0.506755              54   
2  3.128928  4.231226             269   
3 -0.210614 -0.471415              28   
4  3.272894  3.219757             232   

                                  fixed_summary_text  splling_err_num  \
0  The third wave was an experimental see how peo...                5   
1  They would rub it up with soda to make the sme...                2   
2  In Egypt, there were many occupations and soci...               32   
3  The highest class was Pharaohs these people we...                5   
4  The Third Wave developed  rapidly because the ...               29   

                                     prompt_question  \
0  Summarize how the Third Wave developed over su...   
1  Summarize the various ways the factory would u...   
2  In complete sentences, summarize the structure...   
3  In complete sentences, summarize the structure...   
4  Summarize how the Third Wave developed over su...   

                prompt_title  \
0             The Third Wave   
1    Excerpt from The Jungle   
2  Egyptian Social Structure   
3  Egyptian Social Structure   
4             The Third Wave   

                                         prompt_text  prompt_length  \
0  Background \r\nThe Third Wave experiment took ...            660   
1  With one member trimming beef in a cannery, an...           1076   
2  Egyptian society was structured like a pyramid...            625   
3  Egyptian society was structured like a pyramid...            625   
4  Background \r\nThe Third Wave experiment took ...            660   

   length_ratio  word_overlap_count  bigram_overlap_count  \
0      0.096970                  14                     4   
1      0.050186                  18                    22   
2      0.430400                  22                    52   
3      0.044800                   6                     6   
4      0.351515                  23                    27   

   bigram_overlap_ratio  trigram_overlap_count  trigram_overlap_ratio  \
0              0.063492                      0               0.000000   
1              0.415094                     10               0.192308   
2              0.194030                     23               0.086142   
3              0.222222                      5               0.192308   
4              0.116883                      5               0.021739   

   quotes_count  
0             0  
1             0  
2             2  
3             0  
4             4

In [ ]:
import pickle
with open('pickled.pkl', 'wb') as f:
    pickle.dump(train, f)
    pickle.dump(test, f)

In [ ]:
import pickle
with open('pickled.pkl', 'rb') as f:
    train = pickle.load(f)
    test = pickle.load(f)


In [ ]:
# train = train[:512]

In [ ]:
gkf = GroupKFold(n_splits=CFG.n_splits)

for i, (_, val_index) in enumerate(gkf.split(train, groups=train["prompt_id"])):
    train.loc[val_index, "fold"] = i

train.head()

student_id prompt_id                                               text  \
0  000e8c3c7ddb    814d6b  The third wave was an experimentto see how peo...   
1  0020ae56ffbf    ebad26  They would rub it up with soda to make the sme...   
2  004e978e639e    3b9047  In Egypt, there were many occupations and soci...   
3  005ab0199905    3b9047  The highest class was Pharaohs these people we...   
4  0070c9e7af47    814d6b  The Third Wave developed  rapidly because the ...   

    content   wording  summary_length  \
0  0.205683  0.380538              64   
1 -0.548304  0.506755              54   
2  3.128928  4.231226             269   
3 -0.210614 -0.471415              28   
4  3.272894  3.219757             232   

                                  fixed_summary_text  splling_err_num  \
0  The third wave was an experimental see how peo...                5   
1  They would rub it up with soda to make the sme...                2   
2  In Egypt, there were many occupations and soci...               32   
3  The highest class was Pharaohs these people we...                5   
4  The Third Wave developed  rapidly because the ...               29   

                                     prompt_question  \
0  Summarize how the Third Wave developed over su...   
1  Summarize the various ways the factory would u...   
2  In complete sentences, summarize the structure...   
3  In complete sentences, summarize the structure...   
4  Summarize how the Third Wave developed over su...   

                prompt_title  \
0             The Third Wave   
1    Excerpt from The Jungle   
2  Egyptian Social Structure   
3  Egyptian Social Structure   
4             The Third Wave   

                                         prompt_text  prompt_length  \
0  Background \r\nThe Third Wave experiment took ...            660   
1  With one member trimming beef in a cannery, an...           1076   
2  Egyptian society was structured like a pyramid...            625   
3  Egyptian society was structured like a pyramid...            625   
4  Background \r\nThe Third Wave experiment took ...            660   

   length_ratio  word_overlap_count  bigram_overlap_count  \
0      0.096970                  14                     4   
1      0.050186                  18                    22   
2      0.430400                  22                    52   
3      0.044800                   6                     6   
4      0.351515                  23                    27   

   bigram_overlap_ratio  trigram_overlap_count  trigram_overlap_ratio  \
0              0.063492                      0               0.000000   
1              0.415094                     10               0.192308   
2              0.194030                     23               0.086142   
3              0.222222                      5               0.192308   
4              0.116883                      5               0.021739   

   quotes_count  fold  
0             0   3.0  
1             0   2.0  
2             2   1.0  
3             0   1.0  
4             4   3.0

## Model Function Definition

In [ ]:
def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    rmse = mean_squared_error(labels, predictions, squared=False)
    return {"rmse": rmse}

def compute_mcrmse(eval_pred):
    """
    Calculates mean columnwise root mean squared error
    https://www.kaggle.com/competitions/commonlit-evaluate-student-summaries/overview/evaluation
    """
    preds, labels = eval_pred

    col_rmse = np.sqrt(np.mean((preds - labels) ** 2, axis=0))
    mcrmse = np.mean(col_rmse)

    return {
        "content_rmse": col_rmse[0],
        "wording_rmse": col_rmse[1],
        "mcrmse": mcrmse,
    }

def compt_score(content_true, content_pred, wording_true, wording_pred):
    content_score = mean_squared_error(content_true, content_pred)**(1/2)
    wording_score = mean_squared_error(wording_true, wording_pred)**(1/2)

    return (content_score + wording_score)/2

## Deberta Regressor

In [ ]:
class ContentScoreRegressor:
    def __init__(self,
                model_name: str,
                model_dir: str,
                target: str,
                hidden_dropout_prob: float,
                attention_probs_dropout_prob: float,
                max_length: int,
                ):
        self.inputs = ["prompt_text", "prompt_title", "prompt_question", "fixed_summary_text"]
        self.input_col = "input"

        self.text_cols = [self.input_col]
        self.target = target
        self.target_cols = [target]

        self.model_name = model_name
        self.model_dir = model_dir
        self.max_length = max_length

        self.tokenizer = AutoTokenizer.from_pretrained(f"../input/{model_name}")
        self.model_config = AutoConfig.from_pretrained(f"../input/{model_name}")

        self.model_config.update({
            "hidden_dropout_prob": hidden_dropout_prob,
            "attention_probs_dropout_prob": attention_probs_dropout_prob,
            "num_labels": 1,
            "problem_type": "regression",
        })

        seed_everything(seed=42)

        self.data_collator = DataCollatorWithPadding(
            tokenizer=self.tokenizer
        )


    def tokenize_function(self, examples: pd.DataFrame):
        labels = [examples[self.target]]
        tokenized = self.tokenizer(examples[self.input_col],
                         padding=False,
                         truncation=True,
                         max_length=self.max_length)
        return {
            **tokenized,
            "labels": labels,
        }

    def tokenize_function_test(self, examples: pd.DataFrame):
        tokenized = self.tokenizer(examples[self.input_col],
                         padding=False,
                         truncation=True,
                         max_length=self.max_length)
        return tokenized

    def train(self,
            fold: int,
            train_df: pd.DataFrame,
            valid_df: pd.DataFrame,
            batch_size: int,
            learning_rate: float,
            weight_decay: float,
            num_train_epochs: float,
            save_steps: int,
        ) -> None:
        """fine-tuning"""

        sep = self.tokenizer.sep_token
        train_df[self.input_col] = (
                    train_df["prompt_title"] + sep
                    + train_df["prompt_question"] + sep
                    + train_df["fixed_summary_text"]
                  )

        valid_df[self.input_col] = (
                    valid_df["prompt_title"] + sep
                    + valid_df["prompt_question"] + sep
                    + valid_df["fixed_summary_text"]
                  )

        train_df = train_df[[self.input_col] + self.target_cols]
        valid_df = valid_df[[self.input_col] + self.target_cols]

        model_content = AutoModelForSequenceClassification.from_pretrained(
            f"../input/{self.model_name}",
            config=self.model_config,
            ignore_mismatched_sizes=True
        )

        train_dataset = Dataset.from_pandas(train_df, preserve_index=False)
        val_dataset = Dataset.from_pandas(valid_df, preserve_index=False)

        train_tokenized_datasets = train_dataset.map(self.tokenize_function, batched=False)
        val_tokenized_datasets = val_dataset.map(self.tokenize_function, batched=False)

        # eg. "bert/fold_0/"
        model_fold_dir = os.path.join(self.model_dir, str(fold))


        training_args = TrainingArguments(
            output_dir=model_fold_dir,
            load_best_model_at_end=True, # select best model
            learning_rate=learning_rate,
            per_device_train_batch_size=batch_size,
            per_device_eval_batch_size=4,
            num_train_epochs=num_train_epochs,
            weight_decay=weight_decay,
            report_to='none',
            greater_is_better=False,
            save_strategy="steps",
            evaluation_strategy="steps",
            eval_steps=save_steps,
            save_steps=save_steps,
            metric_for_best_model="rmse",
            save_total_limit=1,
            fp16=True,
            warmup_steps = 100,
            gradient_accumulation_steps = 4,
            )

        trainer = Trainer(
            model=model_content,
            args=training_args,
            train_dataset=train_tokenized_datasets,
            eval_dataset=val_tokenized_datasets,
            tokenizer=self.tokenizer,
            compute_metrics=compute_metrics,
            data_collator=self.data_collator
        )

        trainer.train()

        shutil.rmtree(self.model_dir)

        model_content.save_pretrained(self.model_dir)
        self.tokenizer.save_pretrained(self.model_dir)


    def predict(self,
                test_df: pd.DataFrame,
                fold: int,
               ):
        """predict content score"""

        sep = self.tokenizer.sep_token
        in_text = (
                    test_df["prompt_title"] + sep
                    + test_df["prompt_question"] + sep
                    + test_df["fixed_summary_text"]
                  )
        test_df[self.input_col] = in_text

        test_ = test_df[[self.input_col]]

        test_dataset = Dataset.from_pandas(test_, preserve_index=False)
        test_tokenized_dataset = test_dataset.map(self.tokenize_function_test, batched=False)

        model_content = AutoModelForSequenceClassification.from_pretrained(f"{self.model_dir}")
        model_content.eval()

        # e.g. "bert/fold_0/"
        model_fold_dir = os.path.join(self.model_dir, str(fold))

        test_args = TrainingArguments(
            output_dir=model_fold_dir,
            do_train = False,
            do_predict = True,
            per_device_eval_batch_size = 2,
            dataloader_drop_last = False,
        )

        # init trainer
        infer_content = Trainer(
                      model = model_content,
                      tokenizer=self.tokenizer,
                      data_collator=self.data_collator,
                      args = test_args)

        preds = infer_content.predict(test_tokenized_dataset)[0]

        return preds

In [ ]:
def train_by_fold(
        train_df: pd.DataFrame,
        model_name: str,
        target:str,
        save_each_model: bool,
        n_splits: int,
        batch_size: int,
        learning_rate: int,
        hidden_dropout_prob: float,
        attention_probs_dropout_prob: float,
        weight_decay: float,
        num_train_epochs: int,
        save_steps: int,
        max_length:int
    ):

    # delete old model files
    if os.path.exists(model_name):
        shutil.rmtree(model_name)

    os.mkdir(model_name)

    for fold in range(CFG.n_splits):
        print(f"fold {fold}:")

        train_data = train_df[train_df["fold"] != fold]
        valid_data = train_df[train_df["fold"] == fold]

        if save_each_model == True:
            model_dir =  f"{target}/{model_name}/fold_{fold}"
        else:
            model_dir =  f"{model_name}/fold_{fold}"

        csr = ContentScoreRegressor(
            model_name=model_name,
            target=target,
            model_dir = model_dir,
            hidden_dropout_prob=hidden_dropout_prob,
            attention_probs_dropout_prob=attention_probs_dropout_prob,
            max_length=max_length,
           )

        csr.train(
            fold=fold,
            train_df=train_data,
            valid_df=valid_data,
            batch_size=batch_size,
            learning_rate=learning_rate,
            weight_decay=weight_decay,
            num_train_epochs=num_train_epochs,
            save_steps=save_steps,
        )

def validate(
    train_df: pd.DataFrame,
    target:str,
    save_each_model: bool,
    model_name: str,
    hidden_dropout_prob: float,
    attention_probs_dropout_prob: float,
    max_length : int
    ) -> pd.DataFrame:
    """predict oof data"""
    for fold in range(CFG.n_splits):
        print(f"fold {fold}:")

        valid_data = train_df[train_df["fold"] == fold]

        if save_each_model == True:
            model_dir =  f"{target}/{model_name}/fold_{fold}"
        else:
            model_dir =  f"{model_name}/fold_{fold}"

        csr = ContentScoreRegressor(
            model_name=model_name,
            target=target,
            model_dir = model_dir,
            hidden_dropout_prob=hidden_dropout_prob,
            attention_probs_dropout_prob=attention_probs_dropout_prob,
            max_length=max_length,
           )

        pred = csr.predict(
            test_df=valid_data,
            fold=fold
        )

        train_df.loc[valid_data.index, f"{target}_pred"] = pred

    return train_df

def predict(
    test_df: pd.DataFrame,
    target:str,
    save_each_model: bool,
    model_name: str,
    hidden_dropout_prob: float,
    attention_probs_dropout_prob: float,
    max_length : int
    ):
    """predict using mean folds"""

    for fold in range(CFG.n_splits):
        print(f"fold {fold}:")

        if save_each_model == True:
            model_dir =  f"{target}/{model_name}/fold_{fold}"
        else:
            model_dir =  f"{model_name}/fold_{fold}"

        csr = ContentScoreRegressor(
            model_name=model_name,
            target=target,
            model_dir = model_dir,
            hidden_dropout_prob=hidden_dropout_prob,
            attention_probs_dropout_prob=attention_probs_dropout_prob,
            max_length=max_length,
           )

        pred = csr.predict(
            test_df=test_df,
            fold=fold
        )

        test_df[f"{target}_pred_{fold}"] = pred

    test_df[f"{target}"] = test_df[[f"{target}_pred_{fold}" for fold in range(CFG.n_splits)]].mean(axis=1)

    return test_df

In [ ]:
for target in ["content", "wording"]:
    train_by_fold(
        train,
        model_name=CFG.model_name,
        save_each_model=True,
        target=target,
        learning_rate=CFG.learning_rate,
        hidden_dropout_prob=CFG.hidden_dropout_prob,
        attention_probs_dropout_prob=CFG.attention_probs_dropout_prob,
        weight_decay=CFG.weight_decay,
        num_train_epochs=CFG.num_train_epochs,
        n_splits=CFG.n_splits,
        batch_size=CFG.batch_size,
        save_steps=CFG.save_steps,
        max_length=CFG.max_length
    )

fold 0:


fold 1:


fold 2:


fold 3:


fold 0:


fold 1:


fold 2:


fold 3:


In [ ]:
for target in ["content", "wording"]:
    train = validate(
        train,
        target=target,
        save_each_model=True,
        model_name=CFG.model_name,
        hidden_dropout_prob=CFG.hidden_dropout_prob,
        attention_probs_dropout_prob=CFG.attention_probs_dropout_prob,
        max_length=CFG.max_length
    )

    rmse = mean_squared_error(train[target], train[f"{target}_pred"], squared=False)
    print(f"cv {target} rmse: {rmse}")



fold 0:


fold 1:


fold 2:


fold 3:


cv content rmse: 0.7138056605628254
fold 0:


fold 1:


fold 2:


fold 3:


cv wording rmse: 0.8122355945575762


In [ ]:
for target in ["content", "wording"]:
    test = predict(
        test,
        target=target,
        save_each_model=True,
        model_name=CFG.model_name,
        hidden_dropout_prob=CFG.hidden_dropout_prob,
        attention_probs_dropout_prob=CFG.attention_probs_dropout_prob,
        max_length=CFG.max_length
    )

fold 0:


fold 1:


fold 2:


fold 3:


fold 0:


fold 1:


fold 2:


fold 3:


In [ ]:
train.head()

student_id prompt_id                                               text  \
0  000e8c3c7ddb    814d6b  The third wave was an experimentto see how peo...   
1  0020ae56ffbf    ebad26  They would rub it up with soda to make the sme...   
2  004e978e639e    3b9047  In Egypt, there were many occupations and soci...   
3  005ab0199905    3b9047  The highest class was Pharaohs these people we...   
4  0070c9e7af47    814d6b  The Third Wave developed  rapidly because the ...   

    content   wording  summary_length  \
0  0.205683  0.380538              64   
1 -0.548304  0.506755              54   
2  3.128928  4.231226             269   
3 -0.210614 -0.471415              28   
4  3.272894  3.219757             232   

                                  fixed_summary_text  splling_err_num  \
0  The third wave was an experimental see how peo...                5   
1  They would rub it up with soda to make the sme...                2   
2  In Egypt, there were many occupations and soci...               32   
3  The highest class was Pharaohs these people we...                5   
4  The Third Wave developed  rapidly because the ...               29   

                                     prompt_question  \
0  Summarize how the Third Wave developed over su...   
1  Summarize the various ways the factory would u...   
2  In complete sentences, summarize the structure...   
3  In complete sentences, summarize the structure...   
4  Summarize how the Third Wave developed over su...   

                prompt_title  ... length_ratio  word_overlap_count  \
0             The Third Wave  ...     0.096970                  14   
1    Excerpt from The Jungle  ...     0.050186                  18   
2  Egyptian Social Structure  ...     0.430400                  22   
3  Egyptian Social Structure  ...     0.044800                   6   
4             The Third Wave  ...     0.351515                  23   

   bigram_overlap_count  bigram_overlap_ratio  trigram_overlap_count  \
0                     4              0.063492                      0   
1                    22              0.415094                     10   
2                    52              0.194030                     23   
3                     6              0.222222                      5   
4                    27              0.116883                      5   

   trigram_overlap_ratio  quotes_count  fold  content_pred  wording_pred  
0               0.000000             0   3.0      0.145500      0.331895  
1               0.192308             0   2.0     -0.624576     -0.467002  
2               0.086142             2   1.0      0.988445      0.898408  
3               0.192308             0   1.0     -0.222135     -0.396005  
4               0.021739             4   3.0      1.211325      0.442334  

[5 rows x 22 columns]

## upload dataset

In [ ]:
if is_colab:
  import json
  from kaggle.api.kaggle_api_extended import KaggleApi

  def dataset_upload():
      global USERID, EX_NO, UPLOAD_DIR
      id = f'{USERID}/{EX_NO}'
      dataset_metadata = {}
      dataset_metadata['id'] = id
      dataset_metadata['licenses'] = [{'name': 'CC0-1.0'}]
      dataset_metadata['title'] = f'{TITLE}'
      with open(UPLOAD_DIR + 'dataset-metadata.json', 'w') as f:
          json.dump(dataset_metadata, f, indent=4)
      api = KaggleApi()
      api.authenticate()
      # データセットがない場合
      if f'{USERID}/{EX_NO}' not in [str(d) for d in api.dataset_list(user=USERID, search=f'"{EX_NO}"')]:
          api.dataset_create_new(folder=UPLOAD_DIR,
                                convert_to_csv=False,
                                dir_mode='skip'
                                #dir_mode='zip'
                                )
      # データセットがある場合
      else:
          api.dataset_create_version(folder=UPLOAD_DIR,
                                    version_notes='update',
                                    convert_to_csv=False,
                                    delete_old_versions=True,
                                    dir_mode='skip'
                                    #dir_mode='zip'
                                    )

In [ ]:
# %%time
# if is_colab:
#   !mkdir tmp
#   !zip -r ./tmp/output content wording pickled.pkl

In [ ]:
# %%time
# if is_colab:
#   TITLE = 'devert-v3-trained'
#   EX_NO = 'commitlit-2023'  # 実験番号などを入れる、folderのpathにする
#   USERID = 'aruaru0'
#   UPLOAD_DIR = './tmp/'
#   dataset_upload()

## LGBM model

In [ ]:
targets = ["content", "wording"]

drop_columns = ["fold", "student_id", "prompt_id", "text", "fixed_summary_text",
                "prompt_question", "prompt_title",
                "prompt_text"
               ] + targets

In [ ]:
model_dict = {}

for target in targets:
    models = []

    for fold in range(CFG.n_splits):

        X_train_cv = train[train["fold"] != fold].drop(columns=drop_columns)
        y_train_cv = train[train["fold"] != fold][target]

        X_eval_cv = train[train["fold"] == fold].drop(columns=drop_columns)
        y_eval_cv = train[train["fold"] == fold][target]

        dtrain = lgb.Dataset(X_train_cv, label=y_train_cv)
        dval = lgb.Dataset(X_eval_cv, label=y_eval_cv)

        params = {
            'boosting_type': 'gbdt',
            'random_state': 42,
            'objective': 'regression',
            'metric': 'rmse',
            'learning_rate': 0.048,
            'max_depth': 3,
            'lambda_l1': 0.0,
            'lambda_l2': 0.011
        }

        evaluation_results = {}
        model = lgb.train(params,
                          num_boost_round=10000,
                            #categorical_feature = categorical_features,
                          valid_names=['train', 'valid'],
                          train_set=dtrain,
                          valid_sets=dval,
                          callbacks=[
                              lgb.early_stopping(stopping_rounds=30, verbose=True),
                               lgb.log_evaluation(100),
                              lgb.callback.record_evaluation(evaluation_results)
                            ],
                          )
        models.append(model)

    model_dict[target] = models

[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000393 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 816
[LightGBM] [Info] Number of data points in the train set: 353, number of used features: 12
[LightGBM] [Info] Start training from score 0.036049
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Training until validation scores don't improve for 30 rounds
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGB

In [ ]:
import pickle
with open('lgbm.pkl', 'wb') as f:
    pickle.dump(model_dict, f)

In [ ]:
%%time
if is_colab:
  !mkdir tmp
  !zip -r ./tmp/output content wording pickled.pkl lgbm.pkl

  adding: content/ (stored 0%)
  adding: content/another-bert/ (stored 0%)
  adding: content/another-bert/fold_2/ (stored 0%)
  adding: content/another-bert/fold_2/config.json (deflated 53%)
  adding: content/another-bert/fold_2/spm.model (deflated 50%)
  adding: content/another-bert/fold_2/2/ (stored 0%)
  adding: content/another-bert/fold_2/special_tokens_map.json (deflated 54%)
  adding: content/another-bert/fold_2/tokenizer_config.json (deflated 45%)
  adding: content/another-bert/fold_2/tokenizer.json (deflated 77%)
  adding: content/another-bert/fold_2/added_tokens.json (stored 0%)
  adding: content/another-bert/fold_2/pytorch_model.bin (deflated 16%)
  adding: content/another-bert/fold_3/ (stored 0%)
  adding: content/another-bert/fold_3/config.json (deflated 53%)
  adding: content/another-bert/fold_3/spm.model (deflated 50%)
  adding: content/another-bert/fold_3/3/ (stored 0%)
  adding: content/another-bert/fold_3/special_tokens_map.json (deflated 54%)
  adding: content/another

In [ ]:
%%time
if is_colab:
  TITLE = 'devert-v3-trained'
  EX_NO = 'commitlit-2023'  # 実験番号などを入れる、folderのpathにする
  USERID = 'aruaru0'
  UPLOAD_DIR = './tmp/'
  dataset_upload()

Starting upload for file output.zip


100%|██████████| 10.9G/10.9G [02:07<00:00, 91.9MB/s]


Upload successful: output.zip (11GB)
CPU times: user 19.5 s, sys: 16.3 s, total: 35.8 s
Wall time: 2min 10s


## CV Score

In [ ]:
# cv
rmses = []

for target in targets:
    models = model_dict[target]

    preds = []
    trues = []

    for fold, model in enumerate(models):
        X_eval_cv = train[train["fold"] == fold].drop(columns=drop_columns)
        y_eval_cv = train[train["fold"] == fold][target]

        pred = model.predict(X_eval_cv)

        trues.extend(y_eval_cv)
        preds.extend(pred)

    rmse = np.sqrt(mean_squared_error(trues, preds))
    print(f"{target}_rmse : {rmse}")
    rmses = rmses + [rmse]

print(f"mcrmse : {sum(rmses) / len(rmses)}")

content_rmse : 0.5709819936311817
wording_rmse : 0.7539311119629972
mcrmse : 0.6624565527970895


## Predict

In [ ]:
drop_columns = [
                #"fold",
                "student_id", "prompt_id", "text", "fixed_summary_text",
                "prompt_question", "prompt_title",
                "prompt_text",
                "input"
               ] + [
                f"content_pred_{i}" for i in range(CFG.n_splits)
                ] + [
                f"wording_pred_{i}" for i in range(CFG.n_splits)
                ]

In [ ]:
pred_dict = {}
for target in targets:
    models = model_dict[target]
    preds = []

    for fold, model in enumerate(models):
        X_eval_cv = test.drop(columns=drop_columns)

        pred = model.predict(X_eval_cv)
        preds.append(pred)

    pred_dict[target] = preds

In [ ]:
for target in targets:
    preds = pred_dict[target]
    for i, pred in enumerate(preds):
        test[f"{target}_pred_{i}"] = pred

    test[target] = test[[f"{target}_pred_{fold}" for fold in range(CFG.n_splits)]].mean(axis=1)

In [ ]:
test

student_id prompt_id            text  summary_length fixed_summary_text  \
0  000000ffffff    abc123  Example text 1               3     Example text 1   
1  111111eeeeee    def789  Example text 2               3     Example text 2   
2  222222cccccc    abc123  Example text 3               3     Example text 3   
3  333333dddddd    def789  Example text 4               3     Example text 4   

   splling_err_num prompt_question     prompt_title       prompt_text  \
0                0    Summarize...  Example Title 1  Heading\nText...   
1                0    Summarize...  Example Title 2  Heading\nText...   
2                0    Summarize...  Example Title 1  Heading\nText...   
3                0    Summarize...  Example Title 2  Heading\nText...   

   prompt_length  ...  content_pred_0  content_pred_1  content_pred_2  \
0              3  ...       -1.187232       -0.976556       -0.710381   
1              3  ...       -1.187232       -0.976556       -0.710381   
2              3  ...       -1.187232       -0.976556       -0.710381   
3              3  ...       -1.187232       -0.976556       -0.710381   

   content_pred_3   content  wording_pred_0  wording_pred_1 wording_pred_2  \
0       -1.192058 -1.016557       -0.630777       -0.682294      -0.796006   
1       -1.192058 -1.016557       -0.630777       -0.682294      -0.796006   
2       -1.192058 -1.016557       -0.630777       -0.682294      -0.796006   
3       -1.192058 -1.016557       -0.630777       -0.682294      -0.796006   

   wording_pred_3   wording  
0       -1.133059 -0.810534  
1       -1.133059 -0.810534  
2       -1.133059 -0.810534  
3       -1.133059 -0.810534  

[4 rows x 28 columns]

## Create Submission file

In [ ]:
sample_submission

student_id  content  wording
0  000000ffffff      0.0      0.0
1  111111eeeeee      0.0      0.0
2  222222cccccc      0.0      0.0
3  333333dddddd      0.0      0.0

In [ ]:
test[["student_id", "content", "wording"]]

student_id   content   wording
0  000000ffffff -1.016557 -0.810534
1  111111eeeeee -1.016557 -0.810534
2  222222cccccc -1.016557 -0.810534
3  333333dddddd -1.016557 -0.810534

In [ ]:
test[["student_id", "content", "wording"]].to_csv("submission.csv", index=False)

## Summary

CV result is like this.

| | content rmse |wording rmse | mcrmse | LB| |
| -- | -- | -- | -- | -- | -- |
|baseline| 0.494 | 0.630 | 0.562 | 0.509 | [link](https://www.kaggle.com/code/tsunotsuno/debertav3-baseline-content-and-wording-models)|
| use title and question field | 0.476| 0.619 | 0.548 | 0.508 | [link](https://www.kaggle.com/code/tsunotsuno/debertav3-w-prompt-title-question-fields) |
| Debertav3 + LGBM | 0.451 | 0.591 | 0.521 | 0.461 | [link](https://www.kaggle.com/code/tsunotsuno/debertav3-lgbm-with-feature-engineering) |
| Debertav3 + LGBM with spell autocorrect | 0.448 | 0.581 | 0.514 | 0.459 |nogawanogawa's original code
| Debertav3 + LGBM with spell autocorrect and tuning | 0.442 | 0.566 | 0.504 | 0.453 | this notebook |

The CV values improved slightly, and the LB value is improved.

In [ ]:
if is_colab:
  from google.colab import runtime
  runtime.unassign()